run/reloadable.go

```diff
@@ -3,62 +3,164 @@ package run
 import (
 	"os"
 	"os/signal"
+	"sync"
+	"sync/atomic"
 	"syscall"
 
 	"github.com/relex/gotils/logger"
 	"github.com/relex/slog-agent/base"
+	"github.com/relex/slog-agent/defs"
 )
 
-// ReloadableOrchestrator is a wrapper of ordinary Orchestrator(s) to support re-creation of the Orchestrator on config reload
+// sinksByClientNumber is a fix-sized array to hold downstream sinks by client number as array index
+type sinksByClientNumber [base.MaxClientNumber]base.BufferReceiverSink
+
+// addrsByClientNumber is a fix-sized array to hold client addresses by client number as array index
+type addrsByClientNumber [base.MaxClientNumber]string
+
+// ReloadableOrchestrator supports config reload by re-creating the downstream Orchestrator
 //
 // The type is to be paired with Reloader, which provides the function to reload configuration file and create real Orchestrator(s)
 type ReloadableOrchestrator struct {
+	paused          int32
 	downstream      base.Orchestrator        // the real orchestrator
 	renewDownstream func() base.Orchestrator // function to perform reloading and launch a new downstream orchestrator
+	downstreamSinks sinksByClientNumber
+	downstreamAddrs addrsByClientNumber
+	reloadLock      *sync.RWMutex
 }
 
+// NewReloadableOrchestrator creates a reloadable orchestrator wrapping the given downstream orchestrator
 func NewReloadableOrchestrator(downstream base.Orchestrator, renewDownstream func() base.Orchestrator) *ReloadableOrchestrator {
 	// DO NOT use renewDownstream for initial downstream creation, as config reloading is very different from first-time loading
 
+	rorc := &ReloadableOrchestrator{
+		paused:          0,
+		downstream:      downstream,
+		renewDownstream: renewDownstream,
+		downstreamSinks: sinksByClientNumber{},
+		downstreamAddrs: addrsByClientNumber{},
+		reloadLock:      &sync.RWMutex{},
+	}
+
 	// handle signal
-	c := make(chan os.Signal, 1)
+	c := make(chan os.Signal, 100)
```

Comment:
Since we're discarding all messages that appeared while we were reloading the config, we could just use an unbuffered channel:
```suggestion
	c := make(chan os.Signal)
```
See the comment in this example: https://pkg.go.dev/os/signal#example-Notify